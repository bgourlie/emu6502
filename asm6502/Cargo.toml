[package]
name = "asm6502"
version = "0.1.0"
authors = ["W. Brian Gourlie <bgourlie@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
shared6502 = {path = "../shared6502" }
nom = "5.0"
regex = "1"
lazy_static = "1.4.0"
fnv = "1.0.6"

[dev-dependencies]
insta = "0.12.0"