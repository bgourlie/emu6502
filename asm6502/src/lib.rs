mod lexer;
mod parser;
mod resolver;
mod types;

pub use lexer::lex;
pub use parser::parse;
pub use resolver::Resolver;
pub use types::{Span, Token};

#[cfg(test)]
pub fn tlex(input: &'static str) -> Vec<crate::types::Token> {
    let input = crate::Span::new(input);
    let (_, parsed) = crate::lex(input).unwrap();
    parsed.into_iter().map(|(_, token)| token).collect()
}
