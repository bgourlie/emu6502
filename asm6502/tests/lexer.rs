use asm6502::PositionedToken;
use insta::assert_debug_snapshot;
use std::{fs::File, io::Read};

#[test]
fn all_tokens() {
    assert_debug_snapshot!(Fixture::new("all_tokens").tokens());
}

#[test]
fn invalid_tokens() {
    assert_debug_snapshot!(Fixture::new("invalid_tokens").tokens());
}

struct Fixture {
    input: String,
}

impl Fixture {
    fn new(file: &'static str) -> Fixture {
        let mut file = File::open(format!("./tests/fixtures/{}.txt", file)).unwrap();
        let mut input = String::new();
        file.read_to_string(&mut input).unwrap();
        Fixture { input }
    }

    fn tokens(&self) -> Vec<PositionedToken> {
        asm6502::Lexer::new(self.input.as_str()).collect()
    }
}
