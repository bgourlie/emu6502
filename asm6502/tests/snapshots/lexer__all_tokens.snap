---
source: asm6502/tests/lexer.rs
expression: "Fixture::new(\"all_tokens\").tokens()"
---
[
    PositionedToken {
        token: IfStart,
        line: 1,
        span: (
            1,
            3,
        ),
    },
    PositionedToken {
        token: Identifier(
            "ident",
        ),
        line: 1,
        span: (
            4,
            9,
        ),
    },
    PositionedToken {
        token: EqualsOperator,
        line: 1,
        span: (
            10,
            11,
        ),
    },
    PositionedToken {
        token: DecLiteral(
            1,
        ),
        line: 1,
        span: (
            12,
            13,
        ),
    },
    PositionedToken {
        token: Newline,
        line: 1,
        span: (
            13,
            14,
        ),
    },
    PositionedToken {
        token: Mnemonic(
            Brk,
        ),
        line: 2,
        span: (
            5,
            8,
        ),
    },
    PositionedToken {
        token: Newline,
        line: 2,
        span: (
            8,
            9,
        ),
    },
    PositionedToken {
        token: Else,
        line: 3,
        span: (
            1,
            5,
        ),
    },
    PositionedToken {
        token: Newline,
        line: 3,
        span: (
            5,
            6,
        ),
    },
    PositionedToken {
        token: Mnemonic(
            Brk,
        ),
        line: 4,
        span: (
            5,
            8,
        ),
    },
    PositionedToken {
        token: Newline,
        line: 4,
        span: (
            8,
            9,
        ),
    },
    PositionedToken {
        token: IfEnd,
        line: 5,
        span: (
            1,
            6,
        ),
    },
    PositionedToken {
        token: Newline,
        line: 5,
        span: (
            6,
            7,
        ),
    },
]
